libethereum/ExtVM.h

```diff
@@ -97,19 +98,28 @@ class ExtVM : public ExtVMFace
     void suicide(Address _a) final;
 
     /// Return the EVM gas-price schedule for this execution context.
-    EVMSchedule const& evmSchedule() const final
-    {
-        return m_sealEngine.evmSchedule(envInfo().number());
-    }
+    EVMSchedule const& evmSchedule() const final { return m_evmSchedule; }
 
     State const& state() const { return m_s; }
 
     /// Hash of a block if within the last 256 blocks, or h256() otherwise.
     h256 blockHash(u256 _number) final;
 
 private:
+    EVMSchedule const& initEvmSchedule(int64_t _blockNumber, u256 const& _version) const
+    {
+        // Run with a schedule according to hard fork if before Istanbul
+        // Run with a schedule according to account version if Istanbul and later
+        if (_blockNumber < m_sealEngine.chainParams().istanbulForkBlock)
+            return m_sealEngine.evmSchedule(_blockNumber);
+        else
+            return evmScheduleForAccountVersion(_version);
```

Comment:
Wouldn't this work all the time because we have version 0 before Istanbul?

Reply:
Not really, `EvmSchedule` used changes each fork before Istanbul, so this function to work correctly would need to return different ones for version 0 based on block number.
If I understood correctly your question.

Reply:
Yes, you are right.

But I think the more future proof strategy is to select schedule group by version number and then select the schedule within the group by the block number. As I understand the EIP-1702 and as I want to be, we will not bump the version on each HF but only if some breaking changes are introduced.

Reply:
@chfast I see, it makes sense, but also makes the logic of choosing schedule more complicated, as there's not one-to-one relation between schedules and versions.

I changed this just slightly now. It's not exactly as you suggest, but I think more in line with this strategy.
So basically we care about choosing by block number only within the latest version, while all past versions choose the fixed schedule "just before the version bump".
What do you think?